Support multiple `abi` and `platform` values for `pip download`.

katzdm opened the pull request:
As proposed in #6121.

This is useful when fetching wheels for a well-defined interpreter, whose supported ABIs and platforms are known and fixed, different than the one pip is running under.

uranusjr commented:
Would it be better to make the options `action="append"`, instead of inventing a new comma-separated format? IMO it fits pip’s current argument design to do something like this instead:

```
pip download ...
    --abi cp36m --abi cp36 --abi abi3 --abi none
    ...
```

McSinyx commented:
+1 on making it an `append` option, as it won't break backward compatibility.

uranusjr commented:
I don’t think backward compatibility would be an issue either way, since comma is not a valid character in wheel tags anyway 🙂

McSinyx commented:
I was thinking about the name change (e.g. platform -> platforms) but I guess we can keep an alias if we're to go with the comma-separated list.

katzdm commented:
Thanks for the guidance! This is my first would-be contribution to pip, so I'm a bit unfamiliar with internal stylistic conventions - Will look at `action="append"` shortly!

uranusjr commented on src/pip/_internal/cli/cmdoptions.py:
```diff
@@ -494,9 +494,11 @@ def only_binary():
-    help=("Only use wheels compatible with <platform>. "
-          "Defaults to the platform of the running system."),
+    help=("Only use wheels compatible with <platform>. Defaults to the "
+          "platform of the running system. Use multiple options to specify "
```
How about *Use this option multiple times …*?

katzdm replied:
👍Done (for both `--abi` and `--platform`).

uranusjr reviewed (commented):
Sorry for the delay. Just one nitpick on the help message, otherwise this looks good to me. Thanks for the effort!

katzdm commented:
@uranusjr No problem, and thanks for taking a look! I've addressed your feedback, and have added a couple of elementary functional test-cases. Let me know if you have any other points of concern, whether you'd like to see broader test coverage, or whether things look good to you as-is.

Thanks again for your time!

katzdm commented:
@uranusjr Thanks for the review and approval! As this is my first time contributing to `pip`, I'm not familiar with how the review proceeds from here. Are multiple approving reviews typically awaited before merging?

katzdm commented:
Friendly ping 🙂

katzdm commented:
@uranusjr @pradyunsg Wanted to ping this again - I had to rebase and resolve some conflicts yesterday, following other changes that were merged. Would be nice to get this in, to avoid doing so repeatedly.

Still a bit unclear as to whether we're waiting on other steps (e.g. additional reviews) before merging. Some clarification on the usual process would be appreciated.

Thanks in advance!

pradyunsg commented:
Usually, we wait on a couple of reviews or have the primary reviewer be fairly certain that things will work. I'm personally not super comfortable around ABI tags and if/how we should expose them to the end user, so I'm not comfortable reviewing this (and considering the implications of this change) myself. :)

@uranusjr if you reckon this is good to go, I'm happy to merge.

uranusjr commented:
Wheel tags are difficult, but these flags are designed as explicit overrides and the user has always been responsible for keeping the values straight, so I don’t think this PR has much chance to go wrong, actually. I think it’d be fine including this in 20.3.

xavfernandez commented on src/pip/_internal/utils/compatibility_tags.py:
```diff
@@ -130,13 +130,10 @@ def get_supported(
+    if platforms and len(platforms) == 1:
+        # Only expand list of platforms if a single platform was provided.
+        # Otherwise, assume that the list provided is comprehensive.
+        platforms = _get_custom_platforms(platforms[0])
```
Why not call `_get_custom_platforms` on all provided `platforms` ?

I'd find it strange that `pip download --platform manylinux2010 foobar` would successfully find a `manylinux1_x86_64` whel but ` `pip download --platform manylinux2010 --platform manylinux2014 foobar` wouldn't.

katzdm replied:
@xavfernandez I think this is a bit tricky.

I think it's reasonable to assume that anybody providing multiple `--platform` invocations is a fairly expert user, who is targeting a particular python interpreter different than the one through which `pip` is running. Therefore, the user should be the authority on what is and is not supported.

There may be edge-cases in which `manylinux2014` wheels are acceptable, but `manylinux1` wheels are not. For instance, the `manylinux2014` and `manylinux2010` specifications don't allow for binaries to link against the `ncurses` library, whereas `manylinux1` does allow this library. If a user knows that their target environment does not provide `ncurses`, then they might write

```
pip download \
    --platform manylinux2014  --platform manylinux2010 --platform any \
    --abi cp39m --abi abi3 --abi none \
    --python-version 39 \
    --no-deps \
    hiredis==0.3.1
```

To avoid downloading wheels built for `manylinux1`.

This won't be an issue for most users, but it could come up in more complex deployments (e.g. large monorepos), which (I think) is the setting in which this feature is most likely to be useful.

What do you think? We could always add platform-expansion here later - It's probably more difficult to take it back, afterwards.

katzdm replied:
Another question is one of precedence. Given the following two commands:

```
1. pip download --platform manylinux2014_x86_64 --platform linux_x86_64 foobar==0.1.0

2. pip download --platform manylinux_2014_x86_64 --platform any foobar==0.1.0
```

where the available wheels are built for `[linux_x86_64, manylinux2010_x86_64, any]`, what should each command download?

If we perform `manylinux`-expansion, I would expect for `1` to download `linux_86_64` (since I explicitly declared that platform to be acceptable), whereas I would expect for `2` to download `manylinux2010_x86_64` (since `any` should probably be lowest precedence).

On the other hand, one might argue that `manylinux2014_x86_64` should expand "in-place" in the precedence-list, to include all three `manylinux`-platforms. In that case, `1` should download `manylinux2010_x86_64`.

My intention isn't to argue which approach is more "correct" - Instead, it's to point out that the outcome is non-obvious. If we instead tell `pip` to just trust what the user provided, then this question melts away.

uranusjr replied:
I would expects the `--platform` values be simply union-ed, and all expansions done in-place. The wheels downloaded would typically be installed by pip install on a specific platform, which would again filter the wheels. Incompatible wheels downloaded previously wouldn’t really make any practical differences (except some additional local disk space).

katzdm replied:
Fair enough, I suppose I don't feel strongly in either direction.

@uranusjr @xavfernandez I've pushed a change to implement this - Let me know what you think.

uranusjr commented:
I would just union all the specified tags. The wheels downloaded would typically be installed by `pip install` on a specific platform, which would again filter the wheels. Incompatible wheels downloaded previously wouldn’t really make any difference.

uranusjr commented on src/pip/_internal/cli/cmdoptions.py:
```diff
@@ -495,9 +495,11 @@ def only_binary():
     '--platform',
     dest='platform',
```
Nit: Might be a good idea to use the plural form here (same goes for `abi`).

katzdm replied:
Slight disagreement here: The flags will still be named `--platform` and `--abi`, so it feels to me like the names of the fields in `options` should be named likewise. Note that the variables which these fields are then copied to are subsequently named `abis` and `platforms`.

Having says this, let me know if you feel strongly about this, and I'll be happy to make the change.

xavfernandez replied:
I personally agree with @uranusjr.
Cf https://github.com/pypa/pip/blob/ea74bd3d9085234601a0a6199b2ad65c6b6a002f/src/pip/_internal/cli/cmdoptions.py#L395-L396
or https://github.com/pypa/pip/blob/ea74bd3d9085234601a0a6199b2ad65c6b6a002f/src/pip/_internal/cli/cmdoptions.py#L408-L409
or https://github.com/pypa/pip/blob/ea74bd3d9085234601a0a6199b2ad65c6b6a002f/src/pip/_internal/cli/cmdoptions.py#L777-L778
or https://github.com/pypa/pip/blob/ea74bd3d9085234601a0a6199b2ad65c6b6a002f/src/pip/_internal/cli/cmdoptions.py#L790-L791
or most of the options in this file with the action `append` (even if not all)

katzdm replied:
Thanks for the pointers to prior art! Given that it's an established convention, I certainly agree - Just pushed an update.

katzdm commented:
Looks like the MacOS test failures are happening at `HEAD`, and are likely unrelated.

pradyunsg commented:
Yup -- #9030.

xavfernandez commented:
Since #9030 should now be solved, rebasing again (sorry) should make the tests pass.

pradyunsg commented:
Merging given multiple approvals and that @uranusjr wants this in 20.3. :P
